docs/RFCS/20170920_sql_consistency_checks.md

```diff
@@ -0,0 +1,559 @@
+significant factor causing a rise in these types of errors is new
+features that manipulates tables outside of transactions (schema
+changes, backup, CSV import).
+
+In summary, aspects that can be checked include:
+- Secondary indexes have entries for all rows in the table
+- Dangling index references (i.e. missing row but index entry retained)
+- SQL constraints (`CHECK`, `NOT NULL`, `UNIQUE`, `FOREIGN KEY`)
+- Indexes using [STORING][]. Wrong copies of data in the index (either
+  in the index key or the data payload)
+- Tables using [FAMILY][]. Primary index data is organized into column
+  families as per the TableDescriptor (e.g. there is no extraneous
+  column data in a family and there are no extraneous families)
+- Invalid data encodings. Data can't be read from the bytes, or doesn't
+  match expected types.
+- Non-canonical encodings (where we can decode the data, but if we
+  re-encode it we get a different value.)
+- [Composite encodings][] are valid. This includes `DECIMAL` and
+  collated strings types
+- Key encoding correctly reflects the ordering of the decoded values
+
+Schemas also can be checked for the following:
+- Invalid parent ID for interleaved tables
+- Invalid parent ID in general (db doesn't exist anymore)
+- Invalid index definitions (wrong columns)
+- Invalid view descriptors: invalid column types; invalid SQL; invalid
+  number of columns.
+- Invalid column families
+
+[#17626]: https://github.com/cockroachdb/cockroach/issues/17626
+[#17690]: https://github.com/cockroachdb/cockroach/issues/17690
+[#18533]: https://github.com/cockroachdb/cockroach/issues/18533
+[#18705]: https://github.com/cockroachdb/cockroach/issues/18705
+[STORING]: https://www.cockroachlabs.com/docs/stable/create-index.html#store-columns
+[FAMILY]: https://www.cockroachlabs.com/docs/stable/column-families.html
+[Composite encodings]: https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/encoding.md#composite-encoding
+
+# How to use the checks
+
+The checks are separated into a few commands in order to split the
+purpose of them, alongside ones that may require more work than others
+(and will take longer, such as indexes).
+- Checking a secondary index. This checks that the secondary index
+  contains the correct data, and exists no index entries are missing.
+- Checking an SQL constraint. This checks the validity of a table
+  constraint.
+- Checking the table data. This checks that there are no errors in the
+  stored table data, such as encoding errors. i.e. all checks mentioned
+  in "motiviation" except the above two.
+- Checking the database data. An alias for checking all of the tables.
+
+In addition, there will be an option to run all of the checks.
+
+They can be manually invoked via the CLI, with the following statements.
+```sql
+# Checks a secondary index, making sure that all the data stored in it is correct.
+CHECK TABLE <table> INDEX <index_name>;
```

Comment:
```CHECK``` is already a key word that is used. I think we should stick with something arcane for this since it's an arcane and yet useful feature. How about using sql server's DBCC?

Reply:
I'd like the first implementation of this command to just implement

```DBCC TABLE foo.bar INDEX idx```

and nothing else because this is the most pressing validation we'd like to do.

A second implementation can focus on FOREIGN KEYS